node/src/components/storage.rs

```diff
@@ -712,6 +751,37 @@ impl Storage {
             .transpose()
     }
 
+    pub fn get_highest_block_header_for_db_check(&self) -> Option<Block> {
+        let mut txn = self.env.begin_ro_txn().ok()?;
+        if let Some(height) = self.block_height_index.keys().last() {
+            self.get_block_by_height(&mut txn, *height).ok()?
+        } else {
+            None
+        }
+    }
+
+    pub fn get_state_root_hashes_for_trie_check(&self) -> Option<Vec<Blake2bHash>> {
+        let mut blake_hashes: Vec<Blake2bHash> = Vec::new();
+        let header_txn = match self.env.begin_ro_txn() {
+            Ok(txn) => txn,
+            Err(_) => return None,
+        };
+        let mut cursor = match header_txn.open_ro_cursor(self.block_header_db) {
+            Ok(cursor) => cursor,
+            Err(_) => return None,
+        };
+        for (_, raw_val) in cursor.iter() {
+            let header: BlockHeader = lmdb_ext::deserialize(raw_val).ok()?;
+            let blake_hash = Blake2bHash::from(*header.state_root_hash());
+            blake_hashes.push(blake_hash);
+        }
+
+        blake_hashes.sort();
+        blake_hashes.dedup();
```

Comment:
Do we need to `dedup()`? Wouldn't this mean we have duplicate headers with the same hashes in the storage? That should be a fatal error in my opinion.

Reply:
@xcthulhu You would explain this better than myself

Reply:
I would like to see this resolved before merging. /cc @xcthulhu

Reply:
Resolved offline.